dap_aria_mapping/pipeline/semantic_taxonomy/make_taxonomy.py

```diff
@@ -0,0 +1,204 @@
+    fig, axis = plt.subplots(1, num_levels, figsize=(int(num_levels * 8), 8), dpi=300)
+    for idx, cdict in enumerate(cluster_outputs):
+        if not cdict.get("centroid_params", False):
+            axis[idx].scatter(
+                embeddings_2d[:, 0],
+                embeddings_2d[:, 1],
+                c=[e for e in cdict["labels"].values()],
+                s=1,
+            )
+        else:
+            axis[idx].scatter(
+                cdict["centroid_params"]["n_embeddings_2d"][:, 0],
+                cdict["centroid_params"]["n_embeddings_2d"][:, 1],
+                c=cdict["model"][idx].labels_,
+                s=cdict["centroid_params"]["sizes"],
+            )
+        axis[idx].set_title(f"Centroids - Level {idx}")
+    return fig
+
+
+def plot_imbalanced(
+    cluster_outputs: Dict[str, Any],
+    plot_dicts: Sequence[Any],
+    embeddings_2d: np.ndarray,
+) -> plt.Figure:
+    """Plot centroids for each level of clustering.
+
+    Args:
+        cluster_outputs (Tuple[Dict[str, Any], Sequence[Any]]): The output of the clustering
+            function.
+        embeddings_2d (np.ndarray): A two-dimensional embedding of the embeddings.
+
+    Returns:
+        plt.Figure: A figure with a subplot for each level of clustering.
+    """
+    num_levels = len(cluster_outputs[-1]["silhouette"])
+    fig, axis = plt.subplots(1, num_levels, figsize=(int(num_levels * 8), 8), dpi=300)
+
+    for idx, (cdict, cluster) in enumerate(plot_dicts):
+        labels = [int(e) for e in cdict.values()]
+        di = dict(zip(sorted(set(labels)), range(len(set(labels)))))
+        labels = [di[label] for label in labels]
+        _, lvl = divmod(idx, num_levels)
+        make_subplot_embeddings(
+            embeddings=embeddings_2d,
+            clabels=labels,
+            axis=axis.flat[idx],
+            label=f"{cluster[-1]} {str(lvl)}",
+            s=4,
+        )
+    return fig
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--cluster_method",
+        type=str,
+        default="centroids",
+        help="Which cluster config to use",
+    )
+    parser.add_argument(
+        "--plot",
+        action="store_true",
+        help="Whether to plot the results (only centroids config)",
+    )
+    parser.add_argument(
+        "--production",
+        action="store_true",
+        help="Whether to run in production mode",
+    )
+
+    parser.add_argument(
+        "--seed", type=int, default=42, help="Random seed for reproducibility"
+    )
+
+    args = parser.parse_args()
+    np.random.seed(args.seed)
+    random.seed(args.seed)
+
+    if args.production and not any(
+        ["centroids" in args.cluster_method, "imbalanced" in args.cluster_method]
+    ):
+        raise ValueError("Production mode only supports centroids config")
+
+    logger.info("Downloading embeddings from S3")
+    embeddings = get_embeddings()
+    if not args.production:
```

Comment:
I think they use production to mean whether to overwrite the data. When george reviewed my code he told me to use --test for sampling